dot/parachain/backing/can_second.go

```diff
@@ -0,0 +1,141 @@
+// all active leaves.
+//
+// if the candidate can be seconded, returns true and a map of the heads of active leaves to the depths,
+// where the candidate is a member of the fragment tree.
+// Returns false if the candidate cannot be seconded.
+func (cb *CandidateBacking) secondingSanityCheck(
+	hypotheticalCandidate parachaintypes.HypotheticalCandidate,
+	backedInPathOnly bool, //nolint:unparam
+) (bool, map[common.Hash][]uint) {
+	type response struct {
+		depths          []uint
+		head            common.Hash
+		activeLeafState activeLeafState
+	}
+
+	var (
+		responses            []response
+		candidateParaID      parachaintypes.ParaID
+		candidateRelayParent common.Hash
+		membership           = make(map[common.Hash][]uint)
+	)
+
+	switch v := hypotheticalCandidate.(type) {
+	case parachaintypes.HypotheticalCandidateIncomplete:
+		candidateParaID = v.CandidateParaID
+		candidateRelayParent = v.RelayParent
+	case parachaintypes.HypotheticalCandidateComplete:
+		candidateParaID = parachaintypes.ParaID(v.CommittedCandidateReceipt.Descriptor.ParaID)
+		candidateRelayParent = v.CommittedCandidateReceipt.Descriptor.RelayParent
+	}
+
+	for head, leafState := range cb.perLeaf {
+		if leafState.prospectiveParachainsMode.IsEnabled {
+			allowedParentsForPara := cb.implicitView.knownAllowedRelayParentsUnder(head, candidateParaID)
+
+			if !slices.Contains(allowedParentsForPara, candidateRelayParent) {
+				continue
+			}
+
+			responseCh := make(chan parachaintypes.HypotheticalFrontierResponse)
+			cb.SubSystemToOverseer <- parachaintypes.ProspectiveParachainsMessageGetHypotheticalFrontier{
+				HypotheticalFrontierRequest: parachaintypes.HypotheticalFrontierRequest{
+					Candidates:              []parachaintypes.HypotheticalCandidate{hypotheticalCandidate},
+					FragmentTreeRelayParent: &head,
+					BackedInPathOnly:        backedInPathOnly,
+				},
+				Ch: responseCh,
+			}
+
+			res, ok := <-responseCh
+			if ok {
+				var depths []uint
+				for _, val := range res {
+					for _, membership := range val.Memberships {
+						depths = append(depths, membership.Depths...)
+					}
+				}
+				responses = append(responses, response{depths, head, leafState})
+			}
+		} else if head == candidateRelayParent {
+			if bTreeMap, ok := leafState.secondedAtDepth[candidateParaID]; ok {
+				if _, ok := bTreeMap.Get(0); ok {
+					logger.Debug("Refusing to second candidate because leaf is already occupied.")
+					return false, nil
+				}
+			}
+			responses = append(responses, response{[]uint{0}, head, leafState})
+		}
+	}
+
+	if len(responses) == 0 {
+		return false, nil
+	}
+
+	for _, res := range responses {
+		for _, depth := range res.depths {
+			if bTreeMap, ok := res.activeLeafState.secondedAtDepth[candidateParaID]; ok {
+				if _, ok := bTreeMap.Get(depth); ok {
+					logger.Debugf("Refusing to second candidate at depth %d - already occupied.", depth)
+					return false, nil
+				}
+			}
```

Comment:
Lets make these for loops functions then makes it easier to understand these checks and we don't repeat the same check here and in the the `else if` arm

```go
func checkDepthsAgainstLeaftState(depths []uint, leafState activeLeafState, candidateParaID parachaintypes.ParaID) bool {
	for _, depth := range depths {
		if !isSecondedAtDepth(depth, leafState, candidateParaID) {
                         logger.Debugf("Refusing to second candidate at depth %d - already occupied.", depth)
			return false
		}
	}

	return true
}

func isSecondedAtDepth(depth uint, leafState activeLeafState, candidateParaID parachaintypes.ParaID) bool {
	if bTreeMap, ok := leafState.secondedAtDepth[candidateParaID]; ok {
		if _, ok := bTreeMap.Get(depth); ok {
			return false
		}
	}
	return true
}
```